yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,393 @@
+// overriding the --recorder flag.
+func (r *Recorder) SetMode(newMode Mode) {
+	if r.mode == newMode {
+		return
+	}
+	r.mode = newMode
+	r.logger.Logf("recorder %s from/to %v", r.mode.toHumanString(), r.recordsDir)
+}
+
+func sanitizeFilename(s string) (r string) {
+	const allowedRunes = `_-.`
+	return strings.Map(func(rv rune) rune {
+		if unicode.IsLetter(rv) || unicode.IsNumber(rv) {
+			return rv
+		}
+		if strings.ContainsRune(allowedRunes, rv) {
+			return rv
+		}
+		return '_'
+	}, s)
+}
+
+func (r *Recorder) hashRequest(request *transport.Request, body []byte) string {
+	log := r.logger
+	hash := fnv.New64a()
+
+	ha := func(b string) {
+		_, err := hash.Write([]byte(b))
+		if err != nil {
+			log.Fatal(err)
+		}
+		_, err = hash.Write([]byte("."))
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	ha(request.Caller)
+	ha(request.Service)
+	ha(string(request.Encoding))
+	ha(request.Procedure)
+
+	headersMap := request.Headers.Items()
+	orderedHeadersKeys := make([]string, 0, len(headersMap))
+	for k := range headersMap {
+		orderedHeadersKeys = append(orderedHeadersKeys, k)
+	}
+	sort.Strings(orderedHeadersKeys)
+	for _, k := range orderedHeadersKeys {
+		ha(k)
+		ha(headersMap[k])
+	}
+
+	ha(request.ShardKey)
+	ha(request.RoutingKey)
+	ha(request.RoutingDelegate)
+
+	_, err := hash.Write(body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return fmt.Sprintf("%x", hash.Sum64())
+}
+
+func (r *Recorder) makeFilePath(request *transport.Request, hash string) string {
+	s := fmt.Sprintf("%s.%s.%s.yaml", request.Service, request.Procedure, hash)
+	return filepath.Join(r.recordsDir, sanitizeFilename(s))
+}
+
+// Call implements the yarpc transport filter interface
+func (r *Recorder) Call(
+	ctx context.Context,
+	request *transport.Request,
+	out transport.Outbound,
+) (*transport.Response, error) {
+	log := r.logger
+	requestBody, err := ioutil.ReadAll(request.Body)
+	request.Body = bytes.NewReader(requestBody)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestHash := r.hashRequest(request, requestBody)
+	filepath := r.makeFilePath(request, requestHash)
+
+	switch r.mode {
+	case Replay:
+		response, err := r.loadRecord(filepath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		return response, nil
+	case Append:
+		response, err := r.loadRecord(filepath)
+		if err == nil {
+			return response, nil
+		}
+		fallthrough
+	case Overwrite:
+		response, err := out.Call(ctx, request)
+		if err == nil {
+			r.saveRecord(request, requestBody, response, filepath)
+		}
+		return response, err
+	default:
+		panic("Invalid mode")
+	}
+}
+
+// loadRecord attempts to load a record from the given file. If the record
+// cannot be found the errRecordNotFound is returned. Any other error will
+// abort the current test.
+func (r *Recorder) loadRecord(filepath string) (*transport.Response, error) {
+	rawRecord, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errRecordNotFound
```

Comment:
If we want a better error message, either this or the `log.Fatal` in Replay should contain information about the request that was not recorded. Preferably in a readable format (so the `requestRecord` struct rather than the real `Request` struct).